### PR TITLE
Expand OhmmsArray APIs.

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
To prepare for the actual use.

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed

### CODE REVIEW

File: src/Containers/OhmmsPETE/OhmmsArray.h [hunk 1]
```diff
@@ -95,6 +97,74 @@ class Array
...
+  template<typename SIZET = size_t, typename = std::is_integral<SIZET>>
+  Type_t* data(const std::array<SIZET, D>& offsets)
+  {
+    size_t offset = offsets[0];
+    for (int i = 1; i < offsets.size(); i++)
+      offset = offset * Length[i] + offsets[i];
+    return X.data() + offset;
+  }
+  template<typename SIZET = size_t, typename = std::is_integral<SIZET>>
+  const Type_t* data(const std::array<SIZET, D>& offsets) const
```
Comment (PDoakORNL) [inline src/Containers/OhmmsPETE/OhmmsArray.h:109]: As far as I can tell this is an std::array of indexes.
The term offset makes me think these are different.  You can have an offset into a buffer.  And in CUDA at least there is a notion of a 2D and 3D buffer, but I don't think this is that.

File: src/Containers/OhmmsPETE/OhmmsArray.h [hunk 2]
```diff
@@ -95,6 +97,74 @@ class Array
...
+           typename           = qmcplusplus::IsDualSpace<Allocator>>
+  const Type_t* device_data(const std::array<SIZET, D>& offsets) const
+  {
+    size_t offset = offsets[0];
+    for (int i = 1; i < offsets.size(); i++)
+      offset = offset * Length[i] + offsets[i];
+    return X.device_data() + offset;
+  }
+
+  /// access data pointer at (offset_1, ..., offset_D})
```
Comment (PDoakORNL) [inline src/Containers/OhmmsPETE/OhmmsArray.h:139]: Again different from D dimensional indexes?  For docs I'd just put them all in a doxygen group and document there the idea of indexing a D dimensional tensor to a data ptr.

File: src/Containers/OhmmsPETE/OhmmsArray.h [hunk 3]
```diff
@@ -95,6 +97,74 @@ class Array
...
+  template<typename... Args>
+  Type_t* data(Args... offsets)
+  {
+    static_assert(sizeof...(Args) == D, "resize arguments must match dimensionality of Array");
```
Comment (PDoakORNL) [inline src/Containers/OhmmsPETE/OhmmsArray.h:143]: I think the assert message needs to be updated.

File: src/Containers/OhmmsPETE/OhmmsArray.h [hunk 4]
```diff
@@ -95,6 +97,74 @@ class Array
     return X.device_data();
   }
 
+  template<typename SIZET = size_t, typename = std::is_integral<SIZET>>
+  Type_t* data(const std::array<SIZET, D>& offsets)
+  {
+    size_t offset = offsets[0];
+    for (int i = 1; i < offsets.size(); i++)
```
Comment (PDoakORNL) [inline src/Containers/OhmmsPETE/OhmmsArray.h:104]: This is a repetition written differently of the operator() indexing code.  Might be better have it all work the same.  They all calculate a linear index into the data from D indexes.

File: src/Containers/OhmmsPETE/tests/test_Array.cpp [hunk 5]
```diff
@@ -73,6 +73,12 @@ TEST_CASE("array NestedContainers", "[OhmmsPETE]")
   CHECK(vec_copy(0).back() == 123);
 }
 
+TEST_CASE("Array::data", "[OhmmsPETE]")
```
Comment (PDoakORNL) [inline src/Containers/OhmmsPETE/tests/test_Array.cpp:76]: This doesn't seem like quite enough.
Think about what could go wrong when the code gets modified.
Check some edge cases.
Array<T,1>
First and last indicies, 0 and max.
Array where size of the type is different from double.
populate the array and make sure that the dereference is the data at that location (the same error could be easily made in the hand indexing in the test and the data member function.

Basically the same tests can cover the operator() functions.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
Some of the naming and the similar code in operator() functions should be addressed.  The access algorithm seems fine.

File: src/Containers/OhmmsPETE/OhmmsArray.h [hunk 6]
```diff
@@ -95,6 +97,74 @@ class Array
     return X.device_data();
   }
 
+  template<typename SIZET = size_t, typename = std::is_integral<SIZET>>
+  Type_t* data(const std::array<SIZET, D>& offsets)
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/Containers/OhmmsPETE/OhmmsArray.h]: changed to `data_at()`

File: src/Containers/OhmmsPETE/OhmmsArray.h (hunk 2 above)
Comment (ye-luo) [reply src/Containers/OhmmsPETE/OhmmsArray.h]: put `indices` as the argument name

File: src/Containers/OhmmsPETE/OhmmsArray.h (hunk 1 above)
Comment (ye-luo) [reply src/Containers/OhmmsPETE/OhmmsArray.h]: put `indices` as the argument name

File: src/Containers/OhmmsPETE/OhmmsArray.h (hunk 3 above)
Comment (ye-luo) [reply src/Containers/OhmmsPETE/OhmmsArray.h]: Fixed.

File: src/Containers/OhmmsPETE/OhmmsArray.h (hunk 4 above)
Comment (ye-luo) [reply src/Containers/OhmmsPETE/OhmmsArray.h]: Rewrite `operator()` with parameter packing

File: src/Containers/OhmmsPETE/tests/test_Array.cpp (hunk 5 above)
Comment (ye-luo) [reply src/Containers/OhmmsPETE/tests/test_Array.cpp:76]: > Check some edge cases.
Array<T,1>
First and last indicies, 0 and max.
Array where size of the type is different from double.

Not understanding why the type needs to be different from double. Just changed double to float.
Also added value checking.

**Comment by ye-luo** [comment]
Test this please

File: src/Containers/OhmmsPETE/tests/test_Array.cpp (hunk 5 above)
Comment (PDoakORNL) [reply src/Containers/OhmmsPETE/tests/test_Array.cpp:76]: I would say its to make sure future work doesn't somehow only work for double or have some hard coded dependence on the sizeof(double)

**Review by PDoakORNL — APPROVED** [review]
